bundle/jsky.app.ot/src/main/scala/jsky/app/ot/ags/BagsManager.scala

```diff
@@ -1,266 +1,238 @@
+        case _             => // Ignore
       }
-    }
   }
 
-  // Given an observation, either insert it in all data structures or update the existing entry for it.
-  private def updateObservation(obs: ISPObservation) =
-    Option(obs).foreach { ispObs =>
-      val obsKey = ispObs.getNodeKey
-      val ctxOpt = ObsContext.create(ispObs).asScalaOpt
-
-      // Update the task map and the task queue.
-      taskMap.synchronized {
-        val task = taskMap.get(obsKey) match {
-          case Some(t) => t.newContext(ctxOpt)
-          case None    => BagsTask(ispObs, None, ctxOpt)
-        }
-        taskMap += ((obsKey, task))
-      }
-      taskQueue.synchronized {
-        if (!taskQueue.contains(obsKey))
-          taskQueue.add(obsKey)
+  object StructurePropertyChangeListener extends PropertyChangeListener {
+    override def propertyChange(evt: PropertyChangeEvent): Unit =
+      evt.getSource match {
+        case cont: ISPObservationContainer => cont.getAllObservations.asScala.foreach(enqueue(_, 0L))
+        case _                             => // Ignore
       }
-    }
-
-  // Remove a given observation from all data structures.
-  private def removeObservation(obs: ISPObservation) = {
-    Option(obs).foreach { ispObs =>
-      val obsKey = ispObs.getNodeKey
-      taskQueue.synchronized(taskQueue.remove(obsKey))
-      taskMap.synchronized(taskMap -= obsKey)
-    }
   }
+}
 
-  // Check two target environments to see if the BAGS targets match exactly between them.
-  def bagsTargetsMatch(oldEnv: TargetEnvironment, newEnv: TargetEnvironment): Boolean =
-    oldEnv.getGuideEnvironment.getPrimaryReferencedGuiders.equals(newEnv.getGuideEnvironment.getPrimaryReferencedGuiders) &&
-      oldEnv.getGuideEnvironment.getPrimaryReferencedGuiders.asScala.forall { gp =>
-        oldEnv.getPrimaryGuideProbeTargets(gp).asScalaOpt.forall { oldGpt =>
-          val oldBags = oldGpt.getBagsTarget.asScalaOpt
-          val newBags = newEnv.getPrimaryGuideProbeTargets(gp).asScalaOpt.flatMap(_.getBagsTarget.asScalaOpt)
-          (oldBags.isEmpty && newBags.isEmpty) || oldBags.exists(oldTarget => newBags.exists(_.getTarget.equals(oldTarget.getTarget)))
-        }
-      }
+object BagsManager {
+  private val NumThreads = math.max(1, Runtime.getRuntime.availableProcessors - 1)
+  val instance = new BagsManager(new ScheduledThreadPoolExecutor(NumThreads))
 
-  private val compositePropertyChangeListener = new PropertyChangeListener {
-    override def propertyChange(evt: PropertyChangeEvent): Unit = evt.getSource match {
-      case node: ISPNode => updateObservation(node.getContextObservation)
-      case _             => // Ignore
-    }
+
+  // Check two target environments to see if the BAGS targets match exactly between them.
+  def bagsTargetsMatch(oldEnv: TargetEnvironment, newEnv: TargetEnvironment): Boolean = {
+    // Find a group with a BAGS target in it in the old and new envs.
```

Comment:
Of course, I don't know what was discussed with Andy about this but my guess would have been that only the primary guide group matters in this comparison.  In fact I think only the primary guide group should have BAGS targets.  There's no guarantee here that you're even comparing the "same" group since there is no unique identifier for groups.